Tasks done

BaxtiyorNurmatov opened the pull request:
All tasks done. Changes in 1s, 2nd, 3rd commits to tasks respectively

BaxtiyorNurmatov committed:
[d] 1st task done

BaxtiyorNurmatov committed:
[d] 2nd task done

BaxtiyorNurmatov committed:
[d] 3rd task done

antonkazakov commented on app/src/main/java/otus/homework/coroutines/ui/cats/CatsViewModel.kt:
```diff
@@ -0,0 +1,42 @@
+        _catsView = catsView
+    }
+
+    fun detachView() {
```
Я думаю во VM это можно перенести в onCleared, который вызовется сам

antonkazakov commented on app/src/main/java/otus/homework/coroutines/ui/cats/CatsViewModel.kt:
```diff
@@ -0,0 +1,42 @@
+
+    fun fetchCats() {
+        viewModelScope.launch(coroutineExceptionHandler) {
+            val fact = catsService.getCatFact()
```
У тебя на мейнтреде запросы в сеть. Хоть они и неблокирующие лучше их конечно же перевести в другой диспатчер

antonkazakov commented on app/src/main/java/otus/homework/coroutines/ui/cats/CatsPresenter.kt:
```diff
@@ -0,0 +1,42 @@
+
+    fun onInitComplete() = scope.launch {
+        try {
+            val fact = catsService.getCatFact()
```
Нужно переключить диспатчер

antonkazakov commented on app/src/main/java/otus/homework/coroutines/ui/cats/CatsPresenter.kt:
```diff
@@ -0,0 +1,42 @@
+) {
+    private var _catsView: ICatsView? = null
+    private val scope = PresenterScope()
+    private var job: Job? = null
```
У тебя же он всегда null

antonkazakov commented on app/src/main/java/otus/homework/coroutines/ui/cats/CatsPresenter.kt:
```diff
@@ -0,0 +1,42 @@
+
+    fun detachView() {
+        _catsView = null
+        job?.cancel()
```
Отменяй скоуп, а не отдельную джобу

antonkazakov reviewed (changes requested):
Привет. Нужно поправить замечания

BaxtiyorNurmatov committed:
Requested changes done:
1. https://github.com/Otus-Android/Coroutines-Homework/pull/24#discussion_r730246477
2. https://github.com/Otus-Android/Coroutines-Homework/pull/24#discussion_r730246582

antonkazakov commented on app/src/main/java/otus/homework/coroutines/ui/cats/CatsViewModel.kt:
```diff
@@ -0,0 +1,49 @@
+
+    private val coroutineExceptionHandler = CoroutineExceptionHandler { _, throwable ->
+        if (throwable is CancellationException) {
+            throw throwable
```
Здесь уже можно не пробрасывать дальше

antonkazakov commented on app/src/main/java/otus/homework/coroutines/ui/cats/CatsViewModel.kt:
```diff
@@ -0,0 +1,49 @@
+    }
+
+    fun fetchCats() {
+        viewModelScope.launch(coroutineExceptionHandler + Dispatchers.IO) {
```
Тут лучше оставить Main, а async перевести на ИО, хотя повторюсь что они ретрофит коллы будут неблокирующими, так что это опционально

antonkazakov commented on app/src/main/java/otus/homework/coroutines/ui/cats/CatsViewModel.kt:
```diff
@@ -0,0 +1,49 @@
+
+    fun fetchCats() {
+        viewModelScope.launch(coroutineExceptionHandler + Dispatchers.IO) {
+            try {
```
У тебя не будут здесь отлавливаться ошибки, все будут пробрасываться выше. Нужно дать корутинам запущенным через асинк самостоятельно хендлить ошибки, это можно сделать например через функцию `coroutineScope`

antonkazakov reviewed (changes requested):
Спасибо, что поправил замечания. В целом все хорошо, но нужно поправить момент с обработкой ошибок

BaxtiyorNurmatov committed:
Requested changes done:
1. https://github.com/Otus-Android/Coroutines-Homework/pull/24#discussion_r757768239
2. https://github.com/Otus-Android/Coroutines-Homework/pull/24#discussion_r757768341
3. https://github.com/Otus-Android/Coroutines-Homework/pull/24#discussion_r757768404

antonkazakov commented on app/src/main/java/otus/homework/coroutines/ui/cats/CatsViewModel.kt:
```diff
@@ -0,0 +1,48 @@
+    fun fetchCats() {
+        viewModelScope.launch {
+            try {
+                val fact = async(coroutineExceptionHandler + Dispatchers.IO) { catsService.getCatFact() }
```
Я думаю лучше конечно 2 корутины эти объединить в coroutineScope

BaxtiyorNurmatov replied:
А какие плюсы и минусы это?

antonkazakov reviewed (commented):
Оставил замечание. Работу принимаю, но будет круто если поправишь
